### PR TITLE
Components: Ship TypeScript types

### PR DESCRIPTION
**Pull request by mirka**
## What?

Related to #35744

Properly ship first-party TypeScript declarations with the `@wordpress/components` package.

## Why?

We were relying on `@types`, which could be wrong or outdated.

**Follow-up: remove types from Definitely Typed**

## Testing Instructions

1. `npm run build`
2. In your IDE, go to component usages in packages other than wp-components. Check that IntelliSense and source mapping (e.g. "Go to Definition" in VS Code) works as expected.

### CODE REVIEW

**Comment by gziolo** [comment]
Wow, it’s so close 🎉🎉🎉 

That’s quite an achievement, congrats to everyone involved!

**Comment by ciampo** [comment]
I gave this PR a quick look, and types seems to work as expected, at least in VSCode.

It would be good to get more eyes on this, especially:
- from folks working on other parts of Gutenberg
- using other IDEs
- one important thing to check for is any incompatibility with existing Definitely Typed type defs

Pinging a few folks in that regard: @gziolo @aaronrobertshaw @andrewserong @noisysocks @sirreal

File: packages/components/tsconfig.json [hunk 1]
```diff
@@ -43,7 +43,6 @@
 		"src/**/*.native.js",
 		"src/**/react-native-*",
 		"src/**/stories/**/*.js", // only exclude js files, tsx files should be checked
-		"src/**/test/**/*.js", // only exclude js files, ts{x} files should be checked
-		"src/index.js"
```
Comment (ciampo) [inline packages/components/tsconfig.json:47]: On top of excluding `index.js`, I wonder if there's any value in renaming it to `index.ts` ?

**Review by aaronrobertshaw — COMMENTED** [review]
Nice work on this @mirka ✨ 

I've only been able to give this a run using VS Code but it's looking good.

✅ Build process completes without issue
✅ Intellisense tooltips show correct info
✅ Go to source definition lands on the correct file and export

**Review by tyxla — COMMENTED** [review]
I gave this some testing on Sublime Text and it worked well 👍 

* Builds without errors ✅ 
* Tooltips display properly ✅ 
* Going to definition or type definition works as expected ✅

**Review by chad1008 — COMMENTED** [review]
This is really cool... working well for me in VSCode as well.

File: packages/components/tsconfig.json (hunk 1 above)
Comment (ciampo) [reply packages/components/tsconfig.json:47]: I tried renaming it, let's see if anything breaks

**Review by ciampo — APPROVED** [review]
🚀

**Comment by noahtallen** [comment]
DT removal: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65213

**Pull request merged**